e2e/src/helper/MasterAccountManager.ts

```diff
@@ -0,0 +1,405 @@
+        let nonce = startNonce + index;
+        let retries = 0;
+        const maxRetries = 3;
+
+        while (retries < maxRetries) {
+          try {
+            const hash = await walletClient.sendTransaction({
+              to: wallet.address as `0x${string}`,
+              value: parseEther(thenFundWith.toString()),
+              nonce,
+            } as any);
+
+            console.log(`✅ Sent tx ${hash} for ${wallet.address} with nonce ${nonce}`);
+            fundedAddresses.push(wallet.address);
+            return { address: wallet.address, hash, funded: true };
+          } catch (error: any) {
+            if (error.message?.includes('nonce too low') && retries < maxRetries - 1) {
+              retries++;
+              // Get fresh nonce
+              const freshNonce = await publicClient.getTransactionCount({
+                address: masterAccount.address,
+                blockTag: 'pending'
+              });
+              nonce = freshNonce + index;
+              console.log(`🔄 Retrying ${wallet.address} with fresh nonce ${nonce} (attempt ${retries + 1}/${maxRetries})`);
+              await new Promise(resolve => setTimeout(resolve, 500 * retries)); // Exponential backoff
+            } else {
+              console.error(`❌ Failed to fund ${wallet.address} with nonce ${nonce}:`, error.message);
+              return { address: wallet.address, funded: false, error };
+            }
+          }
+        }
+
+        return { address: wallet.address, funded: false, error: new Error('Max retries exceeded') };
+      });
+
+      // Wait for batch to complete
+      await Promise.all(batchPromises);
+
+      // Small delay between batches to ensure transactions are processed
+      if (batchIndex < batches.length - 1) {
+        console.log(`⏳ Waiting 2 seconds before next batch...`);
+        await new Promise(resolve => setTimeout(resolve, 2000));
+      }
+    }
+
+    // Update balances after all funding is complete
+    console.log("📊 Updating balances for funded wallets...");
+    const updatedBalancePromises = fundedAddresses.map(address =>
+      getBalance(publicClient, {
+        address: address as `0x${string}`,
+      }).then(balance => ({
+        address,
+        balance: formatEther(balance)
+      }))
+    );
+
+    const updatedBalances = await Promise.all(updatedBalancePromises);
+
+    // Update the pool with new balances
+    updatedBalances.forEach(({ address, balance }) => {
+      const poolIndex = masterAccountPool.findIndex(poolWallet => poolWallet.address === address);
+      if (poolIndex !== -1) {
+        masterAccountPool[poolIndex].balance = balance;
+        console.log("🔑 Updated balance after funding:", address, balance);
+      }
+    });
+  }
+
+  // Write updated pool back to file
+  fs.writeFileSync('master-account-pool.json', JSON.stringify(masterAccountPool, null, 2));
+};
+
+export const generateAndFundWalletsFromPool = async ({
```

Comment:
similar to above, are we missing a check here that the master account has enough funds to fund the newly generated master account wallets? or were you thinking of doing it optimistically / as much as possible.